### PR TITLE
feat: INSERT and DELETE

### PR DESCRIPTION
**Pull request by Aryan-Rajoria**
Adding Insert and delete statements

### CODE REVIEW

File: eva/optimizer/statement_to_opr_convertor.py [hunk 1]
```diff
@@ -330,13 +330,31 @@ def visit_create_index(self, statement: CreateIndexStatement):
...
+            return None
+
+        self.visit_table_ref(table_ref)
+
+        # Filter Operator
+        predicate = statement.where_clause
+        if predicate is not None:
+            self._visit_select_predicate(predicate)
+
+        if statement.orderby_list is not None:
```
Comment (gaurav274) [inline eva/optimizer/statement_to_opr_convertor.py:345]: Just curious, we shouldn't need orderby/limit in the delete statement, right?

Comment (gaurav274) [reply eva/optimizer/statement_to_opr_convertor.py:345]: Yes, let us start with simple usecase

**Comment by gaurav274** [comment]
Add changes to the https://github.com/georgia-tech-db/eva/blob/master/CHANGELOG.md

File: eva/binder/statement_binder.py [hunk 2]
```diff
@@ -143,6 +144,16 @@ def _bind_select_statement(self, node: SelectStatement):
             self.bind(node.union_link)
             self._binder_context = current_context
 
+    @bind.register(DeleteTableStatement)
+    def _bind_delete_statement(self, node: DeleteTableStatement):
+        self.bind(node.table_ref)
+        if node.where_clause:
+            self.bind(node.where_clause)
+        if node.orderby_list:
```
Comment (gaurav274) [inline eva/binder/statement_binder.py:152]: Remove orderby

File: eva/executor/delete_executor.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+from eva.plan_nodes.project_plan import ProjectPlan
+from eva.utils.logging_manager import logger
+
+
+class DeleteExecutor(AbstractExecutor):
+    """ """
+
+    def __init__(self, node: ProjectPlan):
+        super().__init__(node)
+        self.target_list = node.target_list
```
Comment (gaurav274) [inline eva/executor/delete_executor.py:15]: Remove this

File: eva/executor/delete_executor.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    def validate(self):
+        pass
+
+    def exec(self, **kwargs) -> Iterator[Batch]:
+        # Uses Where clause
+        try:
+            child_executor = self.children[0]
+            for batch in child_executor.exec(*args, **kwargs):
+                batch = apply_predicate(batch, self.predicate)
```
Comment (gaurav274) [inline eva/executor/delete_executor.py:25]: Delete should not be yielding anything. It should rather call `delete` from the storage manager on the table. The predicate should also be based on the delete command.  https://github.com/georgia-tech-db/eva/blob/4b4ee7ac2e01e2eab0d7c17725b36ee011c410f9/eva/storage/sqlite_storage_engine.py#L189
Supporting arbitrary predicates (for example, containing UDFs like similarity ) won't be straightforward. In such cases, we can filter tuples ourselves and then call delete base don `_row_id` but we can think of it as an extension in later PR.

File: eva/executor/insert_executor.py [hunk 5]
```diff
@@ -12,21 +12,81 @@
...
+            table_obj = self.catalog.get_table_catalog_entry(table_name, database_name)
+            
+            # Implemented only for STRUCTURED_DATA
+            if table_obj.table_type != TableType.STRUCTURED_DATA:
+                raise NotImplementedError("INSERT only implemented for structured data")
+            
+            # Values to insert and Columns
+            values = self.node.value_list[0].value
+            column_list = []
+            for i in self.node.column_list:
```
Comment (gaurav274) [inline eva/executor/insert_executor.py:54]: Use better descriptive variable names

File: eva/executor/insert_executor.py [hunk 6]
```diff
@@ -12,21 +12,81 @@
...
+                column_list.append(i.col_name)
+            
+            # Adding all to dataframe
+            dataframe = pd.DataFrame(values, columns=column_list)
+            class Struct(object): pass
```
Comment (gaurav274) [inline eva/executor/insert_executor.py:59]: Didn't understand this? You can consider constructing a batch using `Batch(dataframe)`

File: eva/executor/insert_executor.py [hunk 7]
```diff
@@ -12,21 +12,81 @@
...
+            batch = Struct()
+            batch.frames = dataframe
+            
+            storage_engine = StorageEngine.factory(table_obj)
+            storage_engine.write(table_obj, batch)
+        except Exception as e:
+            if storage_engine and table_obj:
```
Comment (gaurav274) [inline eva/executor/insert_executor.py:66]: No need to rollback as of now.

File: eva/executor/insert_executor.py [hunk 8]
```diff
@@ -12,21 +12,81 @@
...
+                self._rollback_load(storage_engine, table_obj, False)
+            err_msg = f"Load {self.media_type.name} failed: encountered unexpected error {str(e)}"
+            logger.error(err_msg)
+            raise ExecutorError(err_msg)
```
Comment (gaurav274) [inline eva/executor/insert_executor.py:70]: Rewrite this to reflect `insert`

File: eva/optimizer/operators.py [hunk 9]
```diff
@@ -450,6 +454,66 @@ def __hash__(self) -> int:
             )
         )
 
+class LogicalDelete(Operator):
+    """[Logical Node for Delete Operation]
+
+    Arguments:
```
Comment (gaurav274) [inline eva/optimizer/operators.py:460]: Correct the `docstring`

File: eva/optimizer/operators.py [hunk 10]
```diff
@@ -450,6 +454,66 @@ def __hash__(self) -> int:
...
+        table(TableCatalogEntry): table to delete tuples from
+        rows: 
+
+    """
+    def __init__(
+        self,
+        table_ref: TableRef,
+        where_clause: AbstractExpression = None,
+        orderby_clause: AbstractExpression = None,
```
Comment (gaurav274) [inline eva/optimizer/operators.py:469]: Get rid of `orderby_clause` and `limit_clause`

File: eva/optimizer/rules/rules.py [hunk 11]
```diff
@@ -690,6 +692,21 @@ def apply(self, before: LogicalInsert, context: OptimizerContext):
...
+        super().__init__(RuleType.LOGICAL_DELETE_TO_PHYSICAL, pattern)
+
+    def promise(self):
+        return Promise.LOGICAL_DELETE_TO_PHYSICAL
+
+    def check(self, before: Operator, context: OptimizerContext):
+        return True
+
+    def apply(self, before: LogicalDelete, context: OptimizerContext):
+        after = DeletePlan(before.table, before.where_clause, before.orderby_clause, before.limit_count)
```
Comment (gaurav274) [inline eva/optimizer/rules/rules.py:707]: Same as above

File: eva/optimizer/statement_to_opr_convertor.py [hunk 12]
```diff
@@ -316,6 +327,36 @@ def visit_create_index(self, statement: CreateIndexStatement):
...
+        # table_ref = statement.table_ref
+        # if table_ref is None:
+        #     logger.error("From entry missing in select statement")
+        #     return None
+
+        # self.visit_table_ref(table_ref)
+
+        # # Filter Operator
+        # predicate = statement.where_clause
+        # if predicate is not None:
```
Comment (gaurav274) [inline eva/optimizer/statement_to_opr_convertor.py:342]: Remove the commended code

File: eva/optimizer/statement_to_opr_convertor.py [hunk 13]
```diff
@@ -316,6 +327,36 @@ def visit_create_index(self, statement: CreateIndexStatement):
...
+        # if statement.orderby_list is not None:
+        #     self._visit_orderby(statement.orderby_list)
+
+        # if statement.limit_count is not None:
+        #     self._visit_limit(statement.limit_count)
+        
+        delete_opr = LogicalDelete(
+            statement.table_ref,
+            statement.where_clause,
+            statement.orderby_list,
```
Comment (gaurav274) [inline eva/optimizer/statement_to_opr_convertor.py:354]: Same as above

File: eva/parser/delete_statement.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+from eva.parser.statement import AbstractStatement
+from eva.parser.table_ref import TableRef
+from eva.parser.types import StatementType
+
+
+class DeleteTableStatement(AbstractStatement):
+    """
+    Delete Table Statement constructed after parsing the input query
+
+    Attributes
```
Comment (gaurav274) [inline eva/parser/delete_statement.py:28]: We don't use this style of the docstring. Please modify it. Thanks!

File: eva/parser/delete_statement.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        self,
+        table_ref: TableRef,
+        where_clause: AbstractExpression = None,
+        orderby_clause: AbstractExpression = None,
+        limit_count: AbstractExpression = None
+    ):
+        super().__init__(StatementType.DELETE)
+        self._table_ref = table_ref
+        self._where_clause = where_clause
+        self._orderby_list = orderby_clause
```
Comment (gaurav274) [inline eva/parser/delete_statement.py:46]: Same as above

File: eva/parser/delete_statement.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        self._limit_count = limit_count
+
+    def __str__(self) -> str:
+        delete_str = "DELETE FROM {}".format(
```
Comment (gaurav274) [inline eva/parser/delete_statement.py:50]: Try using python `f-strings` as much as possible

File: eva/parser/eva.lark [hunk 17]
```diff
@@ -70,7 +70,7 @@ drop_udf: DROP UDF if_exists? udf_name
 
 //    Primary DML Statements
 
-delete_statement: DELETE FROM table_name (WHERE expression)? order_by_clause? (LIMIT decimal_literal)?
+delete_statement: DELETE FROM table_name (WHERE where_expr)? order_by_clause? (LIMIT decimal_literal)?
```
Comment (gaurav274) [inline eva/parser/eva.lark:73]: Remove `order_by_clause`, `limit`, and `WHERE` can be optional. If the user omits the `where` clause, delete all the rows in the table. We can make omit this corner case if it requires significant changes.  

https://docs.snowflake.com/en/sql-reference/sql/delete.html
https://www.postgresql.org/docs/current/sql-delete.html

File: eva/parser/lark_visitor/_delete_statement.py [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+        where_clause = None
+        order_clause = None
+        limit_count = None
+        for child in tree.children:
+            if isinstance(child, Tree):
+                if child.data == "table_name":
+                    table_name = self.visit(child)
+                    table_ref = TableRef(table_name)
+                elif child.data == "where_expr":
+                    where_clause = self.visit(child)
```
Comment (gaurav274) [inline eva/parser/lark_visitor/_delete_statement.py:37]: Fix this

File: eva/plan_nodes/delete_plan.py [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+        column_list (List[AbstractExpression]): list of annotated column
+        value_list (List[AbstractExpression]): list of abstract expression
+                                                for the values to insert
+    """
+
+    def __init__(
+        self,
+        table_ref: TableRef,
+        where_clause: AbstractExpression = None,
+        orderby_clause: AbstractExpression = None,
```
Comment (gaurav274) [inline eva/plan_nodes/delete_plan.py:38]: Fix this as above

File: test/integration_tests/test_delete_executor.py [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+    def setUp(self):
+        # Bootstrap configuration manager.
+        ConfigurationManager()
+
+        # Reset catalog.
+        CatalogManager().reset()
+
+        load_inbuilt_udfs()
+
+        # Create feature vector table and raw input table.
```
Comment (gaurav274) [inline test/integration_tests/test_delete_executor.py:43]: You can consider using the `create` and `insert` commands to achieve this.

File: test/integration_tests/test_delete_executor.py [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+            )
+        )
+        storage_engine.write(input_tb_entry, input_batch_data)
+
+    def tearDown(self):
+        file_remove("dummy.avi")
+
+    # integration test
+    @unittest.skip("Not supported in current version")
+    def test_should_delete_video_in_table(self):
```
Comment (gaurav274) [inline test/integration_tests/test_delete_executor.py:119]: Rename the test case

File: test/integration_tests/test_delete_executor.py [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+        feat_col_list = [
+            ColumnDefinition("id", ColumnType.INTEGER, NdArrayType.UINT8, (1)),
+            ColumnDefinition("feat", ColumnType.NDARRAY, NdArrayType.FLOAT32, (1, 3)),
+            ColumnDefinition("input", ColumnType.NDARRAY, NdArrayType.UINT8, (1, 3)),
+        ]
+        feat_col_entries = xform_column_definitions_to_catalog_entries(feat_col_list)
+
+        input_col_list = [
+            ColumnDefinition("id", ColumnType.INTEGER, NdArrayType.UINT8, (1)),
+            ColumnDefinition("feat", ColumnType.NDARRAY, NdArrayType.FLOAT32, (1, 3)),
```
Comment (gaurav274) [inline test/integration_tests/test_delete_executor.py:66]: You can consider multiple tables, some with primitive columns and others with ndarray-based columns.

File: test/integration_tests/test_insert_executor.py [hunk 23]
```diff
@@ -36,14 +36,28 @@ def test_should_load_video_in_table(self):
         query = """LOAD VIDEO 'dummy.avi' INTO MyVideo;"""
         execute_query_fetch_all(query)
 
-        insert_query = """ INSERT INTO MyVideo (id, data) VALUES (40,
-                            [[[40, 40, 40] , [40, 40, 40]],
-                            [[40, 40, 40], [40, 40, 40]]]);"""
+        insert_query = """ INSERT INTO MyVideo (id, data) VALUES (
```
Comment (gaurav274) [inline test/integration_tests/test_insert_executor.py:39]: This shouldn't work, as we support inserts to only the structured table.

Comment (gaurav274) [reply test/integration_tests/test_insert_executor.py:39]: Maybe consider the examples required for the labeling application.

**Review by gaurav274 — CHANGES_REQUESTED** [review]

**Comment by gaurav274** [comment]
Add the new commands to the documentation. Thanks!

File: eva/executor/delete_executor.py [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+            elif table_catalog.table_type == TableType.IMAGE_DATA:
+                del_batch =  storage_engine.read(table_catalog)
+            elif table_catalog.table_type == TableType.STRUCTURED_DATA:
+                del_batch = storage_engine.read(table_catalog, batch_mem)
+            
+            del_batch = list(del_batch)[0]
+            original_column_names = list(del_batch.frames.columns)
+            column_names = [f"{table_catalog.name}.{name}" for name in original_column_names if not name=="_row_id"]
+            column_names.insert(0, '_row_id')
+            del_batch.frames.columns = column_names
```
Comment (Aryan-Rajoria) [inline eva/executor/delete_executor.py:58]: Needed to change `column_names` manually, is there a better way to do this?

File: eva/executor/delete_executor.py (hunk 4 above)
Comment (Aryan-Rajoria) [reply eva/executor/delete_executor.py]: Yes Gaurav, thanks for that
I have made changes accordingly.

File: eva/parser/delete_statement.py (hunk 14 above)
Comment (Aryan-Rajoria) [reply eva/parser/delete_statement.py:28]: Is it better now

**Comment by Aryan-Rajoria** [comment]
> Add the new commands to the documentation. Thanks!



> Add changes to the https://github.com/georgia-tech-db/eva/blob/master/CHANGELOG.md

Yes added

**Comment by Aryan-Rajoria** [comment]
> Add the new commands to the documentation. Thanks!

Should I add in different commit?

**Pull request closed without merging**